src/batch/src/executor/row_seq_scan.rs

```diff
@@ -398,22 +398,49 @@ impl<S: StateStore> RowSeqScanExecutor<S> {
             next_col_bounds,
         } = scan_range;
 
-        let (start_bound, end_bound) =
+        let (start_bound, end_bound, null_is_largest) =
             if table.pk_serializer().get_order_types()[pk_prefix.len()].is_ascending() {
-                (next_col_bounds.0, next_col_bounds.1)
+                (next_col_bounds.0, next_col_bounds.1, true)
             } else {
-                (next_col_bounds.1, next_col_bounds.0)
+                (next_col_bounds.1, next_col_bounds.0, false)
             };
 
+        let start_bound_is_bounded = !matches!(start_bound, Bound::Unbounded);
+        let end_bound_is_bounded = !matches!(end_bound, Bound::Unbounded);
+
         // Range Scan.
         assert!(pk_prefix.len() < table.pk_indices().len());
         let iter = table
             .batch_iter_with_pk_bounds(
                 epoch.into(),
                 &pk_prefix,
                 (
-                    start_bound.map(|x| OwnedRow::new(vec![x])),
-                    end_bound.map(|x| OwnedRow::new(vec![x])),
+                    match start_bound {
+                        Bound::Unbounded => {
+                            if end_bound_is_bounded && !null_is_largest {
+                                // Since Null is the smallest value, we need to skip it for range scan to meet the sql semantics.
+                                Bound::Excluded(OwnedRow::new(vec![None]))
+                            } else {
+                                // Both start and end are unbounded, so we need to select all rows.
+                                Bound::Unbounded
+                            }
+                        }
+                        Bound::Included(x) => Bound::Included(OwnedRow::new(vec![x])),
+                        Bound::Excluded(x) => Bound::Excluded(OwnedRow::new(vec![x])),
```

Comment:
Can we also write a test to cover the `DESC` case?

Reply:
Actually, `range_scan.slt.part` itself has already covered this branch, otherwise, our e2e tests will fail.

Reply:
Don't quite get that. 😕 I mean filtering out `NULL` after predicate push-down on a `DESC` key.

Reply:
I mean we have already had these tests.
https://github.com/risingwavelabs/risingwave/blob/962d833946f95b22dfb9b5f9abb7808826fd465f/e2e_test/batch/basic/range_scan.slt.part#L16-L20

https://github.com/risingwavelabs/risingwave/blob/962d833946f95b22dfb9b5f9abb7808826fd465f/e2e_test/batch/basic/range_scan.slt.part#L66-L70

Reply:
`NULL` is always the largest. To cover the issue fixed in this PR, I guess the query should still be `user_id > 43` even though its order is `DESC`. For example, before this PR:

```sql
dev=> create table t (v int primary key);
dev=> create materialized view rev_t as select * from t order by v desc;

dev=> insert into t values (10), (20), (NULL);

dev=> select * from rev_t where v > 15;
 v  
----
 20
   
(2 rows)
```

Reply:
IIRC, NULL in desc order is the smallest value.

Reply:
Anyway, I can add your e2e test case as well.

Reply:
> IIRC, NULL in desc order is the smallest value.

Nope. when `DESC`, NULL should be the **first** item, which is the **largest** item.

Reply:
You can't decide what position `NULL`s are by simply checking `is_ascending`. Everywhere a user can write `ASC/DESC`, he/she can write `NULLS FIRST/LAST`. I submitted a new PR to fix the use of `OrderType` https://github.com/risingwavelabs/risingwave/pull/13567